lib/iris/tests/unit/util/test__slice_data_with_keys.py

```diff
@@ -0,0 +1,257 @@
+from six.moves import (filter, input, map, range, zip)  # noqa
+
+# import iris tests first so that some things can be initialised before
+# importing anything else
+import iris.tests as tests
+
+import numpy as np
+
+from iris.util import _slice_data_with_keys
+
+
+class DummyArray(object):
+    # A dummy array-like that records the keys of indexing calls.
+    def __init__(self, shape, _indexing_record_list=None):
+        self.shape = shape
+        self.ndim = len(shape)
+        if _indexing_record_list is None:
+            _indexing_record_list = []
+        self._getitem_call_keys = _indexing_record_list
+
+    def __getitem__(self, keys):
+        # Add the indexing keys to the call list.
+        self._getitem_call_keys.append(keys)
+        # Return a new object with the correct derived shape, and record its
+        # indexing operations in the same key list as this.
+        shape_array = np.zeros(self.shape)
+        shape_array = shape_array.__getitem__(keys)
+        new_shape = shape_array.shape
+        return DummyArray(new_shape,
+                          _indexing_record_list=self._getitem_call_keys)
+
+
+class Indexer(object):
+    # An object to make __getitem__ arglists from indexing operations.
+    def __getitem__(self, keys):
+        return keys
+
+
+# An Indexer object for generating indexing keys in a nice visible way.
+Index = Indexer()
+
+
+class MixinIndexingTest(object):
+    def check(self, shape, keys, expect_call_keys=None, expect_map=None):
+        data = DummyArray(shape)
+        dim_map, _ = _slice_data_with_keys(data, keys)
+        if expect_call_keys is not None:
+            calls_got = data._getitem_call_keys
+            # Check that the indexing keys applied were the expected ones.
+            equal = len(calls_got) == len(expect_call_keys)
+            for act_call, expect_call in zip(calls_got, expect_call_keys):
+                equal &= len(act_call) == len(expect_call)
+                # A problem here is that in each call, some keys may be
+                # *arrays*, and arrays can't be compared in the "normal"
+                # way.  So we must use np.all for comparison :-(
+                for act_key, expect_key in zip(act_call, expect_call):
+                    equal &= (np.asanyarray(act_key).dtype ==
+                              np.asanyarray(expect_key).dtype and
+                              np.all(act_key == expect_key))
+            errmsg = 'Different key lists:\n{!s}\n!=\n{!s}\n'
+
+            def showkeys(keys_list):
+                msg = '[\n  '
+                msg += '\n  '.join(str(x) for x in keys_list)
+                msg += '\n]'
+                return msg
+
+            self.assertTrue(equal, errmsg.format(showkeys(calls_got),
+                                                 showkeys(expect_call_keys)))
+        if expect_map is not None:
+            self.assertEqual(dim_map, expect_map)
+
+
+class Test_indexing(MixinIndexingTest, tests.IrisTest):
```

Comment:
@pp-mo these tests look good :+1: ... and very thorough!

My only comment, which applies to both test classes here, is that there are very few failure modes. Admittedly there are only two exceptions within all the functions under test here, but are there any other edge cases that need to be considered?